### PR TITLE
12073 product repo get different sku case

### PR DESCRIPTION
**Pull request by josh-carter**
### Description
Take an example Product with SKU `test_sku`

In ProductRepository currently, if we're using get() and pass the SKU `TesT_SKu` we would get an undefined index as described inside #12073.

However It makes sense to be able to retrieve product using different cased SKU as the database is case insensitive and SKU is unique. But the PHP is case sensitive, therefore, to fix the problem we can lowercase the SKU in the PHP code.

### Fixed Issues (if relevant)

1. [magento/magento2#12073: ProductRepository fails to load product with camelcase SKU](https://github.com/magento/magento2/issues/12073)

### Manual testing scenarios

1. Create Product with SKU `test_sku`
2. Make a call to `ProductRepository::get('tESt_sKu')`
3. Successfully retrieve created Product 

### Contribution checklist
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds on Travis CI are green)

### CODE REVIEW

**Comment by orlangur** [comment]
Hi @josh-carter, putting this PR on hold as hope to get fix done in scope of https://github.com/magento/magento2/pull/12169 (slightly different approach: product is cached exactly with the same SKU how it was requested instead of `strtolower` call as it is done in this PR).

**Comment by josh-carter** [comment]
@orlangur Ok cool, my thoughts would be, it saves multiple loading of the same product and keeps functionality correct.

**Review by orlangur — CHANGES_REQUESTED** [review]
Hi @josh-carter, let's continue in this PR :)

Please rework implementation of `cacheProduct` in https://github.com/magento/magento2/commit/cb3b5d1e0715632c15d8a9697538afca4083a859#diff-ff4893a5628d34910295bae32fbcddfdL241 so that code behavior remain unchanged.

Just change implementation of `cacheProduct` in such a way that we pass both `productId` and `sku` instead of obtaining them from product instance.

Another case is that MySQL does not distinguish not just lower/upper case but also е and ё, o and ö, a and ã and so on. Thus there is no need to complicate implementation with `strtolower` or other approaches for key normalization.

**Comment by josh-carter** [comment]
@orlangur Ok that's cool with me. I'll get it updated and get this ready for review.

File: app/code/Magento/Catalog/Model/ProductRepository.php
Comment (orlangur) [inline app/code/Magento/Catalog/Model/ProductRepository.php]: See: https://github.com/magento/magento2/commit/cb3b5d1e0715632c15d8a9697538afca4083a859#diff-ff4893a5628d34910295bae32fbcddfdL241

Please pass `productId` instead of product instance in both cases. Don't add default parameter and some logic into `cacheProduct`, pass both product ID and SKU.

**Review by orlangur — CHANGES_REQUESTED** [review]

File: app/code/Magento/Catalog/Test/Unit/Model/ProductRepositoryTest.php
Comment (orlangur) [inline app/code/Magento/Catalog/Test/Unit/Model/ProductRepositoryTest.php:320]: This test does not seem to test initial case anymore. Please change so that it actually checks trimmed SKU.

File: app/code/Magento/Catalog/Test/Unit/Model/ProductRepositoryTest.php
Comment (orlangur) [inline app/code/Magento/Catalog/Test/Unit/Model/ProductRepositoryTest.php:343]: Same here: it looks like we don't check getting product by `'test_sku'` anymore.

**Review by orlangur — CHANGES_REQUESTED** [review]

**Comment by ibecmattb** [comment]
Facing this same issue. How would I go about patching a Magento 2.2.3 install to fix this?

**Review by ishakhsuvarov — CHANGES_REQUESTED** [review]
I would suggest covering this with different type of test, for instance `api-functional`.

**Comment by sidolov** [comment]
@josh-carter , I am closing this PR now due to inactivity.
Please reopen and update if you wish to continue.
Thank you for collaboration!

**Comment by orlangur** [comment]
Sorry @josh-carter, my bad, I was supposed to help you with test here, I remember we started a talk in Slack. Will do it this week.

**Comment by ishakhsuvarov** [comment]
@orlangur I'm reopening then, to keep track of it

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=14019) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dev.magento.com/magento/magento2?pullRequest=14019) before we can accept your contribution.<br/>

**Review by orlangur — CHANGES_REQUESTED** [review]
Nowadays all changes must be applied to `2.3-develop` first (see https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#rules). Please prepare a new pull request, this one should be `on hold` until the latter is merged.

**Comment by orlangur** [comment]
Closing due to inactivity. Feel free to reach me out anytime later if you wish to continue work on this pull request and it will be reopened.
Thanks for collaboration!

**Pull request closed without merging**
